onnxruntime/core/session/environment.cc

```diff
@@ -250,14 +250,24 @@ Status Environment::Initialize(std::unique_ptr<logging::LoggingManager> logging_
 #ifdef USE_DML
       dml::RegisterDmlSchemas();
 #endif
-      RegisterOnnxOperatorSetSchema();
+
+      // Call once to trigger schema static registration.
+      // Note: This is not transactional and can still collide with concurrent calls from other libs.
+      if (!OpSchemaRegistry::Schema("Add")) {
```

Comment:
Is `__ONNX_DISABLE_STATIC_REGISTRATION` set by default so that this call doesn't trigger registration of all schemas?

What's the reason we need to do this at the ORT level vs. ONNX gracefully handling multiple calls to schema registration? The check for an semi-arbitrary operator name feels a little hacky.

Reply:
> Is `__ONNX_DISABLE_STATIC_REGISTRATION` set by default so that this call doesn't trigger registration of all schemas?

`ONNX_DISABLE_STATIC_REGISTRATION ` is off by default, AFAIK it's just a backdoor for whoever need to load a subset of opset manually, for memory or collision avoidence.
- https://github.com/onnx/onnx/pull/3266

> What's the reason we need to do this at the ORT level vs. ONNX gracefully handling multiple calls to schema registration?

You'll need to ask them for sure.
But clearly the static/shared protobuf discussion didn't move much forward in the past few years.
And google themselves is always doing that monorepo + everything-static thing, so I'm not expecting this to end well in a foreseeable future.

> The check for an semi-arbitrary operator name feels a little hacky.

Yes, so I picked the very basic Add.

Reply:
If you specify `ONNX_DISABLE_STATIC_REGISTRATION` when creating the shared onnx library you're linking against, is the change in ORT required? 

I'm trying to understand why we need to add this hack to ORT given it's not necessary in the typical usage of ORT.

Reply:
That would be a breaking change to onnx behavior as no op would be registered, making onnx itself unusable in normal cases.